### PR TITLE
Reconfiguring ML-Wrappers Wrapping 

### PR DESCRIPTION
**Pull request by natalie-isak**
This PR moves the custom drise wrapper into ml-wrapper repo. 

The purpose of this is to ensure configurability and simplicity when calling image ml wrappers. With this change, all image ml wrapper classes are located in a single file.

### CODE REVIEW

File: python/ml_wrappers/model/image_model_wrapper.py [hunk 1]
```diff
@@ -15,6 +15,8 @@
 from ml_wrappers.model.pytorch_wrapper import WrappedPytorchModel
 from ml_wrappers.model.wrapped_classification_model import \
     WrappedClassificationModel
+from torchvision import transforms as T
```
Comment (Advitya17) [inline python/ml_wrappers/model/image_model_wrapper.py:18]: Seems we're taking a dependency on `torchvision`. Should this be included in the dependencies file, or (more preferably) we should instead mention somewhere in README that torchvision may be needed and hence download your own torch & torchvision module versions from pytorch's official website.

Comment (Advitya17) [reply python/ml_wrappers/model/image_model_wrapper.py:18]: Update: Looks like torchvision was already used in this repo before. Suggest following repo's principles on imports not directly included in dependencies.

```suggestion
try:
    from torchvision import transforms as T
except ImportError:
    module_logger.debug('Could not import torchvision, required if using' +
                        ' a vision PyTorch model')
```

File: python/ml_wrappers/model/image_model_wrapper.py [hunk 2]
```diff
@@ -53,7 +49,7 @@ def _is_fastai_model(model):
     return str(type(model)).endswith(FASTAI_MODEL_SUFFIX)
 
 
-def _wrap_image_model(model, examples, model_task, is_function):
+def _wrap_image_model(model, examples, model_task, is_function, num_classes=None):
```
Comment (Advitya17) [inline python/ml_wrappers/model/image_model_wrapper.py:52]: Consider same arg name between `num_classes` in this function and `number_of_classes` in `WrappedObjectDetectionModel`. If they have different meaning, can perhaps tweak naming and/or clarify in the docstrings.

File: python/ml_wrappers/model/image_model_wrapper.py [hunk 3]
```diff
@@ -249,53 +248,93 @@ def predict_proba(self, dataset: pd.DataFrame) -> np.ndarray:
...
+                # We approximate the score for each class
+                # by dividing (class score) evenly among the other classes.
+
+                raw_detection = filter_score(raw_detection)
+                # to do - add in expanded class scores
+                # expanded_class_scores = od_common.expand_class_scores(
+                #     raw_detection['scores'],
+                #     raw_detection['labels'],
+                #     self._number_of_classes)
+                image_predictions = torch.cat((raw_detection["labels"]
```
Comment (Advitya17) [inline python/ml_wrappers/model/image_model_wrapper.py:329]: Consider pasting the example label representation from the TDD to a comment above this, to explain the formatted output of the torch operations.

File: python/ml_wrappers/model/image_model_wrapper.py [hunk 4]
```diff
@@ -249,53 +248,93 @@ def predict_proba(self, dataset: pd.DataFrame) -> np.ndarray:
...
+            for raw_detection in raw_detections:
+                raw_detection = apply_nms(raw_detection)
+
+                # Note that FasterRCNN doesn't return a score for each class, only
+                # the predicted class. DRISE requires a score for each class.
+                # We approximate the score for each class
+                # by dividing (class score) evenly among the other classes.
+
+                raw_detection = filter_score(raw_detection)
+                # to do - add in expanded class scores
```
Comment (Advitya17) [inline python/ml_wrappers/model/image_model_wrapper.py:324]:
```suggestion
                # TODO - add in expanded class scores
```
  Nit: `TODO` is highlighted with a different color in IDEs like VS code (blue) :D.

File: python/ml_wrappers/model/image_model_wrapper.py [hunk 5]
```diff
@@ -249,53 +248,93 @@ def predict_proba(self, dataset: pd.DataFrame) -> np.ndarray:
...
+                :return: Model predictions filtered out by score_thresh
+                :rtype: dict
+                """
+                keep = orig_prediction['scores'] > score_thresh
+
+                filter_prediction = orig_prediction
+                filter_prediction['boxes'] = filter_prediction['boxes'][keep]
+                filter_prediction['scores'] = filter_prediction['scores'][keep]
+                filter_prediction['labels'] = filter_prediction['labels'][keep]
+                return filter_prediction
```
Comment (Advitya17) [inline python/ml_wrappers/model/image_model_wrapper.py:313]: Consider pushing `apply_nms` and `filter_score` outside `predict`'s definition and invoke `self.{function}` instead.

File: python/ml_wrappers/model/image_model_wrapper.py [hunk 6]
```diff
@@ -307,5 +346,100 @@ def predict_proba(self, dataset, iou_threshold=0.1):
         type iou_threshold: float
         """
         predictions = self.predict(dataset, iou_threshold)
-        prob_scores = [[pred[-1] for pred in image_prediction] for image_prediction in predictions]
+        prob_scores = [[pred.class_scores for pred in image_prediction] for image_prediction in predictions]
         return prob_scores
+
+
+class PytorchFasterRCNNWrapper(
```
Comment (Advitya17) [inline python/ml_wrappers/model/image_model_wrapper.py:353]:
```suggestion
class PytorchDRiseWrapper(
```

Comment (Advitya17) [reply python/ml_wrappers/model/image_model_wrapper.py:353]: Can accordingly tweak docstrings as well.

File: python/ml_wrappers/model/image_model_wrapper.py [hunk 7]
```diff
@@ -307,5 +346,100 @@ def predict_proba(self, dataset, iou_threshold=0.1):
...
+            :return: Model predictions filtered out by score_thresh
+            :rtype: dict
+            """
+            keep = orig_prediction['scores'] > score_thresh
+
+            filter_prediction = orig_prediction
+            filter_prediction['boxes'] = filter_prediction['boxes'][keep]
+            filter_prediction['scores'] = filter_prediction['scores'][keep]
+            filter_prediction['labels'] = filter_prediction['labels'][keep]
+            return filter_prediction
```
Comment (Advitya17) [inline python/ml_wrappers/model/image_model_wrapper.py:419]: If both `apply_nms` and `filter_score` have the same function definitions in the 2 wrappers, consider having an interface wrapper that defines these functions, and have the two wrappers inherit the functions off of that interface and define their own `predict` & `predict_proba`.

File: python/ml_wrappers/model/image_model_wrapper.py [hunk 8]
```diff
@@ -15,6 +15,8 @@
 from ml_wrappers.model.pytorch_wrapper import WrappedPytorchModel
 from ml_wrappers.model.wrapped_classification_model import \
     WrappedClassificationModel
+from torchvision import transforms as T
+from vision_explanation_methods.explanations import common as od_common
```
Comment (Advitya17) [inline python/ml_wrappers/model/image_model_wrapper.py:19]: Curious if we should indeed have this as required dependency and/or trap the ImportError here as well. CC @imatiach-msft

File: tests/main/test_image_model_wrapper.py [hunk 9]
```diff
@@ -19,9 +19,12 @@
                                  retrieve_or_train_fridge_model)
 from ml_wrappers import wrap_model
 from ml_wrappers.common.constants import ModelTask
+from ml_wrappers.model.image_model_wrapper import PytorchFasterRCNNWrapper
+from torchvision import transforms as T
```
Comment (Advitya17) [inline tests/main/test_image_model_wrapper.py:23]: Similar trapping for import error

**Review by Advitya17 — CHANGES_REQUESTED** [review]

File: python/ml_wrappers/model/image_model_wrapper.py (hunk 1 above)
Comment (natalie-isak) [reply python/ml_wrappers/model/image_model_wrapper.py:18]: Good catch!! Thanks, will change

File: python/ml_wrappers/model/image_model_wrapper.py (hunk 3 above)
Comment (natalie-isak) [reply python/ml_wrappers/model/image_model_wrapper.py:328]: Great idea!!! Could use the reminder myself :) Adding

File: python/ml_wrappers/model/image_model_wrapper.py [hunk 10]
```diff
@@ -249,53 +287,75 @@ def predict_proba(self, dataset: pd.DataFrame) -> np.ndarray:
...
+                raw_detections = self._model(
+                    T.ToTensor()(image).to(self._device).unsqueeze(0))
+
+            for raw_detection in raw_detections:
+                raw_detection = _apply_nms(raw_detection)
+
+                # Note that FasterRCNN doesn't return a score for each class,
+                # only the predicted class. DRISE requires a score for each.
+                # We approximate the score for each class
+                # by dividing (class score) evenly among the other classes.
```
Comment (Advitya17) [inline python/ml_wrappers/model/image_model_wrapper.py:342]: What does this mean? Just wanted to know for my understanding.

File: python/ml_wrappers/model/image_model_wrapper.py [hunk 11]
```diff
@@ -249,53 +287,75 @@ def predict_proba(self, dataset: pd.DataFrame) -> np.ndarray:
...
+
+                raw_detection = _filter_score(raw_detection)
+                # TODO - add in expanded class scores
```
Comment (Advitya17) [inline python/ml_wrappers/model/image_model_wrapper.py:345]: 1. What's the utility of this?
2. Adding TODOs is fine, curious if we should have code in comments or remove it as per @imatiach-msft's previous suggestion that we should very rarely have commented code if logic, documentation etc may change and if that code snippet could be rendered incorrect.

**Review by Advitya17 — APPROVED** [review]

File: requirements-automl.txt [hunk 12]
```diff
@@ -1,3 +1,4 @@
 mlflow
 azureml-automl-core
-azureml-automl-dnn-vision
\ No newline at end of file
+azureml-automl-dnn-vision
+vision_explanation_methods
```
[the remark this answers is not in the document]
Comment (natalie-isak) [reply requirements-automl.txt:4]: Just checked, it is <<import vision_explanation_methods>>

(name of package is vision-explanation-methods, though)

File: python/ml_wrappers/model/image_model_wrapper.py [hunk 13]
```diff
@@ -19,11 +19,12 @@
...
 
-
 try:
     import torch
     import torch.nn as nn
     import torchvision
+    from torchvision import transforms as T
+    from vision_explanation_methods.explanations import common as od_common
 except ImportError:
     module_logger.debug('Could not import torch, required if using a PyTorch model')
```
Comment (imatiach-msft) [inline python/ml_wrappers/model/image_model_wrapper.py:29]: nit: the error message here is "Could not import torch, required if using a PyTorch model", but this can be thrown with this PR if:
vision_explanation_methods package import failed
torchvision package import failed
so I think it might be better to move those into their own try/except sections?

**Review by imatiach-msft — APPROVED** [review]

File: python/ml_wrappers/model/image_model_wrapper.py [hunk 14]
```diff
@@ -307,5 +356,62 @@ def predict_proba(self, dataset, iou_threshold=0.1):
...
+    any other models with the same output class.
+
+    :param model: Object detection model
+    :type model: PytorchFasterRCNN model
+    :param number_of_classes: Number of classes the model is predicting
+    :type number_of_classes: int
+    """
+
+    def __init__(self, model, number_of_classes: int):
+        """Initialize the PytorchDRiseWrapper."""
```
Comment (imatiach-msft) [inline python/ml_wrappers/model/image_model_wrapper.py:380]: nit: missing docs for params

File: python/ml_wrappers/model/image_model_wrapper.py [hunk 15]
```diff
@@ -53,7 +60,47 @@ def _is_fastai_model(model):
...
+    :type orig_prediction: dict
+    :param score_thresh: Score threshold to filter by
+    :type score_thresh: float
+    :return: Model predictions filtered out by score_thresh
+    :rtype: dict
+    """
+    keep = orig_prediction['scores'] > score_thresh
+
+    filter_prediction = orig_prediction
+    filter_prediction['boxes'] = filter_prediction['boxes'][keep]
```
Comment (imatiach-msft) [inline python/ml_wrappers/model/image_model_wrapper.py:76]: nit: move repeating strings to top of file as constants, including:

```
'boxes'
'scores'
'labels'
```

you can set them at top of file as:

```
BOXES = 'boxes'
LABELS = 'labels'
SCORS = 'scores'
```

File: python/ml_wrappers/model/image_model_wrapper.py [hunk 16]
```diff
@@ -249,53 +299,64 @@ def predict_proba(self, dataset: pd.DataFrame) -> np.ndarray:
...
-    def __init__(self, model: Any) -> None:
+    def __init__(self, model: Any, number_of_classes: int) -> None:
         """Initialize the WrappedObjectDetectionModel with the model and evaluation function.
 
         :param model: mlflow model
         :type model: Any
         """
-        self._device = torch.device("cuda" if torch.cuda.is_available() else "cpu")
-
+        self._device = torch.device("cuda" if torch.cuda.is_available()
```
[the remark this answers is not in the document]
Comment (Advitya17) [reply python/ml_wrappers/model/image_model_wrapper.py:311]: Device has now been added to the entrypoint `wrap_model` function, so resolving this comment.
